spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java

```diff
@@ -39,28 +48,122 @@
 	HttpTracing httpTracing(
-			@Value("${spring.sleuth.http.legacy.enabled:false}") boolean legacyEnabled,
 			Tracing tracing,
-			TraceKeys traceKeys,
-			ErrorParser errorParser,
 			SkipPatternProvider provider
 	) {
-		if (legacyEnabled) {
-			return HttpTracing.newBuilder(tracing)
-					.clientParser(new SleuthHttpClientParser(traceKeys))
-					.serverParser(new SleuthHttpServerParser(traceKeys, errorParser))
-					.serverSampler(new SleuthHttpSampler(provider))
-					.build();
-		}
-		return HttpTracing
-				.newBuilder(tracing)
-				.serverSampler(new SleuthHttpSampler(provider))
+		return HttpTracing.newBuilder(tracing)
+				.clientParser(new CompositeClientParser(this.clientParsers))
+				.serverParser(new CompositeServerParser(this.serverParsers))
+				.clientSampler(this.clientSampler)
+				.serverSampler(new CompositeServerSampler(this.serverSampler, provider))
 				.build();
 	}
+
+	@Bean
+	@ConditionalOnProperty(name = "spring.sleuth.http.legacy.enabled", havingValue = "true")
+	HttpClientParser sleuthHttpClientParser(TraceKeys traceKeys) {
+		return new SleuthHttpClientParser(traceKeys);
+	}
+
+	@Bean
+	@ConditionalOnProperty(name = "spring.sleuth.http.legacy.enabled", havingValue = "true")
+	HttpServerParser sleuthHttpServerParser(TraceKeys traceKeys, ErrorParser errorParser) {
+		return new SleuthHttpServerParser(traceKeys, errorParser);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(name = "sleuthClientSampler")
+	HttpSampler sleuthClientSampler() {
+		return HttpSampler.TRACE_ID;
+	}
+}
+
+class CompositeClientParser extends HttpClientParser {
+
+	private final List<HttpClientParser> parsers;
+
+	CompositeClientParser(List<HttpClientParser> parsers) {
+		this.parsers = parsers;
+		if (this.parsers.isEmpty()) {
+			this.parsers.add(new HttpClientParser());
+		}
+	}
+
+	@Override public <Req> void request(HttpAdapter<Req, ?> adapter, Req req,
+			SpanCustomizer customizer) {
+		this.parsers.forEach(parser -> parser.request(adapter, req, customizer));
+	}
+
+	@Override public <Resp> void response(HttpAdapter<?, Resp> adapter, Resp res,
+			Throwable error, SpanCustomizer customizer) {
+		this.parsers.forEach(parser -> parser.response(adapter, res, error, customizer));
+	}
+}
+
+class CompositeServerParser extends HttpServerParser {
+
+	private final List<HttpServerParser> parsers;
+
+	CompositeServerParser(List<HttpServerParser> parsers) {
+		this.parsers = parsers;
+		if (this.parsers.isEmpty()) {
+			this.parsers.add(new HttpServerParser());
+		}
+	}
+
+	@Override public <Req> void request(HttpAdapter<Req, ?> adapter, Req req,
+			SpanCustomizer customizer) {
+		this.parsers.forEach(parser -> parser.request(adapter, req, customizer));
+	}
+
+	@Override public <Resp> void response(HttpAdapter<?, Resp> adapter, Resp res,
+			Throwable error, SpanCustomizer customizer) {
+		this.parsers.forEach(parser -> parser.response(adapter, res, error, customizer));
+	}
 }
+
+class CompositeServerSampler extends HttpSampler {
```

Comment:
Personally I think the skipPattern sampler is fantastically easy to override. Just have folks use the property as opposed to composing over it?

Ex SkipPattern is public so someone can easily include it in their own sampler.:
```java
class MyHttpSampler extends HttpSampler {

	private final Pattern pattern;

	SleuthHttpSampler(SkipPatternProvider provider) {
		this.pattern = provider.skipPattern();
	}

	@Override public <Req> Boolean trySample(HttpAdapter<Req, ?> adapter, Req request) {
		String url = adapter.path(request);
		boolean shouldSkip = this.pattern.matcher(url).matches();
		if (shouldSkip) {
			return false;
		}
		return null;
	}
}
```